src/darwin/functions.cpp

```diff
@@ -38,115 +39,36 @@ MOUNTUTILS_RESULT translate_dissenter(DADissenterRef dissenter) {
   }
 }
 
-void unmount_callback(DADiskRef disk, DADissenterRef dissenter, void *context) {
-  MountUtilsLog("Unmount callback called");
-  unmount_done = true;
-  CFRunLoopRef loop = (CFRunLoopRef)context;
-
-  if (dissenter != NULL) {
-    MountUtilsLog("Dissenter returned");
-    code = translate_dissenter(dissenter);
-  }
-
-  MountUtilsLog("Stopping run loop from callback");
-  CFRunLoopStop(loop);
-}
-
-MOUNTUTILS_RESULT unmount_whole_disk(const char *device) {
-  MountUtilsLog("Creating session");
-
-  DASessionRef session = DASessionCreate(kCFAllocatorDefault);
-  if (session == NULL) {
-    MountUtilsLog("Session couldn't be created");
-    return MOUNTUTILS_ERROR_GENERAL;
-  }
-
-  MountUtilsLog("Starting run loop");
-  CFRunLoopRef loop = CFRunLoopGetCurrent();
-  DASessionScheduleWithRunLoop(session, loop, kCFRunLoopDefaultMode);
-  DADiskRef disk = DADiskCreateFromBSDName(kCFAllocatorDefault,
-                                           session,
-                                           device);
-
-  MountUtilsLog("Unmounting disk");
-
-  DADiskUnmount(disk,
-                kDADiskUnmountOptionWhole | kDADiskUnmountOptionForce,
-                unmount_callback,
-                reinterpret_cast<void *>(loop));
-
-  if (!unmount_done) {
-    MountUtilsLog("Stopping run loop");
-    CFRunLoopRun();
-    DASessionUnscheduleFromRunLoop(session, loop, kCFRunLoopDefaultMode);
-    CFRelease(session);
-  }
-
-  return code;
-}
-
-void _eject_cb(DADiskRef disk, DADissenterRef dissenter, void *context) {
-  MountUtilsLog("[eject]: Eject callback");
-  if (dissenter) {
-    MountUtilsLog("[eject]: Eject dissenter");
-    code = translate_dissenter(dissenter);
-  } else {
-    MountUtilsLog("[eject]: Eject success");
-    code = MOUNTUTILS_SUCCESS;
-  }
-  CFRunLoopRef run_loop = (CFRunLoopRef)context;
-  CFRunLoopStop(run_loop);
-}
-
-void _unmount_cb(DADiskRef disk, DADissenterRef dissenter, void *context) {
-  MountUtilsLog("[eject]: Unmount callback");
-  CFRunLoopRef run_loop = (CFRunLoopRef) context;
-  if (dissenter) {
-    MountUtilsLog("[eject]: Unmount dissenter");
-    code = translate_dissenter(dissenter);
-    CFRunLoopRef run_loop = (CFRunLoopRef)context;
-    CFRunLoopStop(run_loop);
-  } else {
-    MountUtilsLog("[eject]: Unmount success");
-    MountUtilsLog("[eject]: Ejecting...");
-    DADiskEject(disk,
-      kDADiskEjectOptionDefault,
-      _eject_cb,
-      reinterpret_cast<void *>(run_loop));
-  }
-}
-
-MOUNTUTILS_RESULT eject_disk(const char* device) {
-  // Reset the result code
-  code = MOUNTUTILS_SUCCESS;
+MOUNTUTILS_RESULT run_cb(const char* device, DADiskUnmountCallback callback) {
+  RunLoopContext *context = new RunLoopContext;
```

Comment:
Any reason why we're allocating this in the heap, given it only needs to be alive during this function (and can be passed by reference to any other function that makes use of it)?